accepted/2023/simplify-workload-versioning.md

```diff
@@ -0,0 +1,194 @@
+When updating workloads, console output will include the old and new workload set versions:
+
+```
+> dotnet workload install wasm-tools
+Checking for updated workload set...
+Updating workload set version from 8.0.201-baseline.23919.2 to 8.0.201...
+Installing workload manifest microsoft.net.sdk.android version 34.0.0-preview.4.230...
+<Further workload manifest update messages>
+Installing pack Microsoft.NET.Runtime.WebAssembly.Sdk version 8.0.0-preview.4.23181.9...
+<Further workload pack installation manifests>
+Garbage collecting for SDK feature band(s) 8.0.200...
+
+Successfully updated workload set version from 8.0.201-baseline.23919.2 to 8.0.201.
+Successfully installed workload(s) wasm-tools.
+```
+
+The proposed [workload history](https://github.com/dotnet/sdk/pull/30486) command will also display the workload set versions.
+
+The .NET SDK will periodically (once a day, by default) check for updated workload versions in order to notify users if there is an update available.  Currently, commands such as `dotnet build` print out the following message:
+
+```
+Workload updates are available. Run `dotnet workload list` for more information.
+```
+
+The `dotnet workload list` command, in turn, prints out the following message:
+
+```
+Updates are available for the following workload(s): wasm-tools. Run `dotnet workload update` to get the latest.
+```
+
+We will modify both commands to print the same message:
+
+```
+A workload update (version 8.0.202) is available. Run `dotnet workload update` to install the latest workloads.
+```
+
+We will add a new `--version` option to `dotnet workload update` to allow installing (or downgrading to) a specific workload set version:
+
+```
+> dotnet workload update --version 8.0.202
+Updating workload set version from 8.0.201 to 8.0.202...
+<Manifest and pack installation / garbage collection messages>
+Successfully updated workload set version from 8.0.201 to 8.0.202.
+```
+
+## Specifying workload versions with global.json
+
+We will add support for specifying the workload set version in global.json.  For example:
+
+```json
+{
+  "sdk": {
+    "workloadVersion": "8.0.201"
+  }
+}
+```
+
+This would force the SDK to use workload set version `8.0.201`, and would error if that version was not installed.
+
+NOTE: We may not implement the global.json workload set version support in the same release as the rest of the changes described in this design.
+
+## Side by side workload sets
+
+The workload `restore`, `update`, and `install` commands will be updated to check if there is a global.json file specifying a workload set version applying to the current folder.  If so, these commands will install that workload set version side-by-side with any other installed workload set.  They will also record the path to the global.json file and the workload set version it specified.  This record will be used as a root during garbage collection to prevent the side-by-side workload set (and its packs) from being removed.
```

Comment:
Does this mean some file in `/usr/local/share/dotnet` will include a path to `/Users/aklig/work/fooProject/global.json` ?

What if I rename the `fooProject` folder to `barProject`?

Reply:
Maybe it would be better to have an explicit `sudo dotnet workload pin <workload version set>` and `sudo dotnet workload unpin <workload version set>` command that could protect a workload set from being removed by `dotnet workload clean` ?   And then `dotnet workload --info` would display if a certain workload set is pinned or not.

This is what Homebrew does when you need to prevent some out of date version from being removed.